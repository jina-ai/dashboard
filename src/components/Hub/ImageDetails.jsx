import React from "react";
import { Card, CardHeader, CardBody, Row, Col } from "shards-react";

export default function ImageDetails({ image }) {
  return (
    <Card className="readme-container mb-4">
      <CardHeader className="border-bottom d-flex flex-row">
        <h6 className="m-0 d-inline-block">Image Details</h6>
      </CardHeader>
      <CardBody className="pt-0 pb-2">
        <Row className="border-bottom py-2">
          <Col>Name</Col>
          <Col>{image.name}</Col>
        </Row>
        <Row className="border-bottom py-2">
          <Col>Version</Col>
          <Col>{image.version}</Col>
        </Row>
        <Row className="border-bottom py-2">
          <Col>Author</Col>
          <Col>{image.author}</Col>
        </Row>
        <Row className="border-bottom py-2">
          <Col>Description</Col>
          <Col>{image.description}</Col>
        </Row>
        <Row className="border-bottom py-2">
          <Col>Documentation</Col>
          <Col>
            <a
              href={image.documentation}
              target="_blank"
              rel="noopener noreferrer"
            >
              {image.documentation}
            </a>
          </Col>
        </Row>
        <Row className="border-bottom py-2">
          <Col>Source</Col>
          <Col>
            <a href={image.source} target="_blank" rel="noopener noreferrer">
              {image.source}
            </a>
          </Col>
        </Row>
        <Row className="border-bottom py-2">
          <Col>License</Col>
          <Col>{image.license}</Col>
        </Row>
        <Row className="border-bottom py-2">
          <Col>Vendor</Col>
          <Col>{image.vendor}</Col>
        </Row>
        <Row className="py-2">
          <Col>URL</Col>
          <Col>
            <a href={image.url} target="_blank" rel="noopener noreferrer">
              {image.url}
            </a>
          </Col>
        </Row>
      </CardBody>
    </Card>
  );
}
