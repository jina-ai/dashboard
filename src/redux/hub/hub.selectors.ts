import { State } from "../index"

// Todo: Ideally, we shouldn't remove keywords. This should be fixed with providing correct data
const tagsToHide = [
  "Some keywords to describe the executor",
  "separated by commas",
]

export const selectHubImages = (state: State) =>
  state.hubState.images.map((image) => {
    const keywords = image.keywords
    return {
      ...image,
      keywords: keywords.filter((keyword) => !tagsToHide.includes(keyword)),
    }
  })

export const selectIsHubImagesLoading = (state: State) => state.hubState.loading

export const selectHubImagesFetchError = (state: State) => state.hubState.error
