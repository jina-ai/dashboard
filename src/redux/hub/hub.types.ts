import {
  FETCH_HUB_IMAGES,
  FETCH_HUB_IMAGES_SUCCESS,
  FETCH_HUB_IMAGES_FAILURE,
} from "./hub.constants"

export type HubImage = {
  "docker-name": string
  version: string
  "jina-version": string
  "docker-command": string
  name: string
  description: string
  type: string
  kind: string
  keywords: string[]
  platform: string[]
  license: string
  url: string
  documentation: string
  author: string
  avatar: string | null
}

export type HubState = {
  images: HubImage[]
  loading: boolean
  error: Error | null
}

export type FetchHubImagesAction = {
  type: typeof FETCH_HUB_IMAGES
}

export type FetchHubImagesSuccessAction = {
  type: typeof FETCH_HUB_IMAGES_SUCCESS
  payload: {
    images: HubImage[]
  }
}

export type FetchHubImagesFailureAction = {
  type: typeof FETCH_HUB_IMAGES_FAILURE
  payload: {
    error: Error
  }
}
export type HubActionTypes =
  | FetchHubImagesAction
  | FetchHubImagesSuccessAction
  | FetchHubImagesFailureAction
