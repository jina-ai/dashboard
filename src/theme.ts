const PERSIAN_GREEN = "#009999"
const GOLDEN_TANOI = "#FFCC66"
const BITTER_SWEET = "#ff6666"
const SCOOTER_GREEN = "#32C8CD"
const WHITE = "#FFFFFF"
const DADGREY = "#DAD7FE"
const SKY_DIVING_IN_THE_BLUE = "#007AFF"
const QUICK_SILVER = "#A6A6A6"
const HONEYBEE = "#FFCC66"
const BLACK = "#000000"

const theme = {
  palette: {
    background: WHITE,
    headerTextColor: BLACK,
    primary: PERSIAN_GREEN,
    secondary: GOLDEN_TANOI,
    warning: GOLDEN_TANOI,
    error: BITTER_SWEET,
    success: SCOOTER_GREEN,
    info: PERSIAN_GREEN,
    critical: "#ff4540",
    debug: "#7E7278",
    tagBackground: DADGREY,
    tagPlatformBackground: HONEYBEE,
    mutedText: QUICK_SILVER,
    highlight: SKY_DIVING_IN_THE_BLUE,
  },
}

export { theme }
