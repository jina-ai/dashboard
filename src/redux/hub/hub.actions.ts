import {
  FETCH_HUB_IMAGES,
  FETCH_HUB_IMAGES_SUCCESS,
  FETCH_HUB_IMAGES_FAILURE,
} from "./hub.constants"
import {
  HubActionTypes,
  HubImage,
  FetchHubImagesSuccessAction,
  FetchHubImagesFailureAction,
} from "./hub.types"
import { Dispatch } from "redux"
import { getHubImages } from "../../services/hubApi"
import { FilterParams } from "../../components/Hub/HubFilters"
import { AppThunk } from ".."

const defaultParams = { kind: [], keywords: [] }

export const fetchHubImages = (
  filters: FilterParams = defaultParams
): AppThunk<Promise<void>> => async (dispatch: Dispatch<HubActionTypes>) => {
  try {
    dispatch({
      type: FETCH_HUB_IMAGES,
    })

    const images = await getHubImages(filters)

    dispatch(fetchHubImagesSuccess(images))
  } catch (e) {
    dispatch(fetchHubImagesFailure(e))
  }
}

const fetchHubImagesSuccess = (
  images: HubImage[]
): FetchHubImagesSuccessAction => {
  return {
    type: FETCH_HUB_IMAGES_SUCCESS,
    payload: { images },
  }
}

const fetchHubImagesFailure = (error: Error): FetchHubImagesFailureAction => {
  return {
    type: FETCH_HUB_IMAGES_FAILURE,
    payload: { error },
  }
}
