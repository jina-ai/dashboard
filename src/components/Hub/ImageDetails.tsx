import React from "react"
import { Card, CardBody } from "shards-react"
import { HubImage } from "../../redux/hub/hub.types"
import styled from "@emotion/styled"
import { Tag } from "./ImageCard"

type Props = {
  image: HubImage
}

const SectionHeading = styled.div`
  font-size: 1.25rem;
  font-weight: 700;
  color: ${(props) => props.theme.palette.headerTextColor};
`
const DetailsSection = styled.div`
  display: flex;
  flex-direction: column;
`
const Platform = styled(Tag)`
  background-color: ${(props) => props.theme.palette.tagPlatformBackground};
`

export default function ImageDetails({ image }: Props) {
  let { keywords, platform, author } = image
  let dockerCommand = image["docker-command"]
  return (
    <Card className="readme-container mb-4">
      <CardBody className="pt-0 pb-2">
        <DetailsSection>
          <SectionHeading>Tags</SectionHeading>
          <div>
            {keywords.map((keyword) => (
              <Tag data-name="hubImageTags" key={keyword}>
                {keyword}
              </Tag>
            ))}
          </div>
        </DetailsSection>
        <DetailsSection>
          <SectionHeading>Platform</SectionHeading>
          <div>
            {platform.map((p) => (
              <Platform data-name="hubImageTags" key={p}>
                {p}
              </Platform>
            ))}
          </div>
        </DetailsSection>
        <DetailsSection>
          <SectionHeading>Docker Pull Command</SectionHeading>
          <div>{dockerCommand}</div>
        </DetailsSection>
        <DetailsSection>
          <SectionHeading>Owner</SectionHeading>
          <div>{author}</div>
        </DetailsSection>
      </CardBody>
    </Card>
  )
}
