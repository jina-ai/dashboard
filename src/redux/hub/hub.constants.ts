import { HubState } from "./hub.types"

export const FETCH_HUB_IMAGES = "FETCH_HUB_IMAGES"
export const FETCH_HUB_IMAGES_SUCCESS = "FETCH_HUB_IMAGES_SUCCESS"
export const FETCH_HUB_IMAGES_FAILURE = "FETCH_HUB_IMAGES_FAILURE"

export const HIDE_BANNER_TIMEOUT = 5000

export const initialHubState: HubState = {
  images: [],
  loading: false,
  error: null,
}
