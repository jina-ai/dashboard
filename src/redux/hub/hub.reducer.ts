import { HubState, HubActionTypes } from "./hub.types"
import {
  FETCH_HUB_IMAGES,
  FETCH_HUB_IMAGES_SUCCESS,
  FETCH_HUB_IMAGES_FAILURE,
  initialHubState,
} from "./hub.constants"
import produce from "immer"

const hubReducer = produce((draft: HubState, action: HubActionTypes) => {
  switch (action.type) {
    case FETCH_HUB_IMAGES:
      draft.loading = true
      break
    case FETCH_HUB_IMAGES_SUCCESS:
      draft.loading = false
      draft.images = action.payload.images
      break
    case FETCH_HUB_IMAGES_FAILURE:
      draft.loading = false
      draft.error = action.payload.error
      break
  }
}, initialHubState)

export default hubReducer
