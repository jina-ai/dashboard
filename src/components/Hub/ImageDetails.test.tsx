import React from "react"
import renderer from "react-test-renderer"
import { ThemeProvider } from "@emotion/react"
import { theme } from "../../theme"
import ImageDetails from "./ImageDetails"
import { HubImage } from "../../redux/hub/hub.types"

test("ImageDetails", () => {
  const image = {
    platform: ["Enel"],
    author: "Oda",
    keywords: ["encoder"],
    "docker-command": "whatever",
  }
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <ImageDetails image={(image as unknown) as HubImage} />
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
