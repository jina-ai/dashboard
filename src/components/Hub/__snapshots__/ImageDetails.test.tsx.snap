// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageDetails 1`] = `
<div
  className="readme-container mb-4 card"
>
  <div
    className="pt-0 pb-2 card-body"
  >
    <div
      className="css-1iyoj2o"
    >
      <div
        className="css-1kjfrvx"
      >
        Tags
      </div>
      <div>
        <div
          className="css-1gs0ali"
          data-name="hubImageTags"
        >
          encoder
        </div>
      </div>
    </div>
    <div
      className="css-1iyoj2o"
    >
      <div
        className="css-1kjfrvx"
      >
        Platform
      </div>
      <div>
        <div
          className="css-14vr8vv"
          data-name="hubImageTags"
        >
          Enel
        </div>
      </div>
    </div>
    <div
      className="css-1iyoj2o"
    >
      <div
        className="css-1kjfrvx"
      >
        Docker Pull Command
      </div>
      <div>
        whatever
      </div>
    </div>
    <div
      className="css-1iyoj2o"
    >
      <div
        className="css-1kjfrvx"
      >
        Owner
      </div>
      <div>
        Oda
      </div>
    </div>
  </div>
</div>
`;
